unit_test.py

```diff
@@ -0,0 +1,1371 @@
+
+
+# generate a list of random value candidates(the number of which is random)
+# in required format
+def rand_value_candidates(min_number=1, max_number=10):
+    candidates = randlength_word()
+    for _ in range(randint(min_number-1, max_number-1)):
+        candidates += (randlength_word()+',')
+    return candidates
+
+
+# generate a random lab id in the required format
+def rand_lab_id():
+    return randlength_word()+':'+randlength_word()
+
+
+# generate a random experiment id in the required format
+def rand_experiment_id():
+    return rand_lab_id()+':'+randlength_word()
+
+
+# generate a random value for classTime attribute in Class class
+def rand_classtime(classtime_list=['16fall', '17spring', '17fall']):
+    return classtime_list[randint(0, len(classtime_list)-1)]
+
+
+def rand_classid():
+    return randlength_word()+'_'+rand_classtime()
+
+
+# generate a power lists
+def rand_powers():
+    m.Role.insert_roles()
+    permission_list = [m.Permission.DATA_ENTRY,
+                       m.Permission.DATA_EDIT,
+                       m.Permission.LAB_SETUP,
+                       m.Permission.ADMIN,
+                       m.Permission.LAB_MANAGE,
+                       m.Permission.USER_MANAGE,
+                       m.Permission.SUPERADMIN]
+    shuffle(permission_list)
+    return [u.get_power(p) for p in
+            permission_list[0:randint(1, len(permission_list))]]
+
+
+# Generate a random value for role_name attribute in User class
+def rand_role(roles=['SuperAdmin', 'Admin', 'Student']):
+    return roles[randint(0, len(roles)-1)]
+
+
+# Generate a random value for role_name attribute in User class
+# among all the possible admin role names
+def rand_admin(roles=['SuperAdmin', 'Admin']):
+    return roles[randint(0, len(roles)-1)]
+
+
+# Generate a random integer which is the number of a loop in many testing
+# functions
+def rand_round(min_round=1, max_round=5):
+    return randint(min_round, max_round)
+
+
+# Generate a random list of students in string format with comma as delimitor
+def rand_student_names(number=5):
+    s = ''
+    for _ in range(number):
+        username = randlength_word()
+        # Avoid repetition of username
+        while s.find(username) != -1:
+            username = randlength_word()
+        s += (username+',')
+    s = s.rstrip(',')
+    return s
+
+
+class TestCase(unittest.TestCase):
+    # this is automatically called for us when we run the test
+    def setUp(self):
+        # It disables the error catching during request handling so that we get
+        # better error reports when performing test requests against the
+        # application.
+        app.config['TESTING'] = True
+        # The database used for this suit of tests is not the one used by the
+        # app. Before we run this test, we need to create a local database
+        # called lemur_test
+        app.config['SQLALCHEMY_DATABASE_URI'] = 'postgres://localhost/lemur_test'
```

Comment:
you should use https://docs.python.org/3/library/unittest.mock.html to mock out database interactions. It *is* useful to have tests that talk to the db, but they are called integration or functional tests, and they often end up being too expensive to run each time you make a small change. (your unit test suite, which should run 100% in memory, or maybe use an occasional temp file, should be fast and cheap unless your app grows truly large)